datajunction-server/datajunction_server/construction/build_v2.py

```diff
@@ -179,202 +184,398 @@ async def get_measures_query(  # pylint: disable=too-many-locals
+            ),
+            ctes=[*node_ast.ctes, node_ast],
+        )
+        node_ast.ctes = []
+        node_ast = node_ast.to_cte(node_alias)
+        cte_mapping[self.node_revision.name] = node_ast
+
+        # Start building the dimension joins and adding them to the CTEs
+        for _, dimension_join in dimension_node_joins.items():
+            join_path = dimension_join.join_path
+            requested_dimensions = list(
+                dict.fromkeys(dimension_join.requested_dimensions),
             )
-            # Add it to the list of CTEs
-            cte_mapping[link.dimension.name] = dimension_join.node_query  # type: ignore
-            final_ast.ctes.append(dimension_join.node_query)  # type: ignore
-
-            # Build the join statement
-            join_ast = build_join_for_link(
-                link,
-                cte_mapping,
-                dimension_node_query,
+
+            for link in join_path:
+                link = cast(DimensionLink, link)
+                if all(
+                    dim in link.foreign_keys_reversed for dim in requested_dimensions
+                ):  # pylint: disable=line-too-long # pragma: no cover
+                    continue  # pragma: no cover
+
+                if link.dimension.name in cte_mapping:
+                    dimension_join.node_query = cte_mapping[link.dimension.name]
+                    continue
+
+                dimension_node_query = await build_dimension_node_query(
+                    self.session,
+                    self._build_criteria,
+                    link,
+                    self._filters,
+                    cte_mapping,
+                )
+                dimension_join.node_query = convert_to_cte(
+                    dimension_node_query,
+                    final_ast,
+                    link.dimension.name,
+                )
+                # Add it to the list of CTEs
+                cte_mapping[link.dimension.name] = dimension_join.node_query  # type: ignore
+                final_ast.ctes.append(dimension_join.node_query)  # type: ignore
+
+                # Build the join statement
+                join_ast = build_join_for_link(
+                    link,
+                    cte_mapping,
+                    dimension_node_query,
+                )
+                final_ast.select.from_.relations[-1].extensions.append(join_ast)  # type: ignore
+
+            # Add the requested dimensions to the final SELECT
+            if join_path:  # pragma: no cover
+                dimensions_columns = build_requested_dimensions_columns(
+                    requested_dimensions,
+                    join_path[-1],
+                    dimension_node_joins,
+                )
+                final_ast.select.projection.extend(dimensions_columns)
+
+        # Mark any remaining requested dimensions that don't need a join with
+        # their canonical dimension names
+        for dim_name in self.dimensions:
```

Comment:
Same as my other comment, I think this can be wrapped into a build step method called something like `self.set_canonical_dimension_aliases()`...or a better named method :)